src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -84,25 +94,69 @@ private StreamingClientProvider() {
+    SnowflakeStreamingIngestClient resultClient = null;
+
     if (Boolean.parseBoolean(
         connectorConfig.getOrDefault(
             SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG,
             Boolean.toString(ENABLE_STREAMING_CLIENT_OPTIMIZATION_DEFAULT)))) {
-      LOGGER.info(
-          "Streaming client optimization is enabled, returning the existing streaming client if"
-              + " valid");
-      this.providerLock.lock();
-      // recreate streaming client if needed
-      if (!StreamingClientHandler.isClientValid(this.parameterEnabledClient)) {
-        LOGGER.error("Current streaming client is invalid, recreating client");
-        this.parameterEnabledClient = this.streamingClientHandler.createClient(connectorConfig);
+      LOGGER.debug(
+          "Streaming client optimization is enabled per worker node. Reusing valid clients when"
+              + " possible");
+
+      try {
+        this.providerLock.lock();
+
+        // remove invalid clients
+        List<Map.Entry<Properties, SnowflakeStreamingIngestClient>> invalidClientProps =
+            this.registeredClientMap.entrySet().stream()
+                .filter(entry -> !StreamingClientHandler.isClientValid(entry.getValue()))
+                .collect(Collectors.toList());
+        if (!invalidClientProps.isEmpty()) {
+          String invalidClientNames = "";
+
+          for (Map.Entry<Properties, SnowflakeStreamingIngestClient> entry : invalidClientProps) {
+            SnowflakeStreamingIngestClient invalidClient = entry.getValue();
+            invalidClientNames +=
+                invalidClient != null
+                        && invalidClient.getName() != null
+                        && !invalidClient.getName().isEmpty()
+                    ? invalidClient.getName()
+                    : "noClientNameFound";
+            this.registeredClientMap.remove(entry.getKey());
+          }
+
+          LOGGER.error(
+              "Found and removed {} invalid clients: {}",
+              invalidClientProps.size(),
+              invalidClientNames);
+        }
+
+        // look for client corresponding to the input properties or create new client
+        Properties inputProps = StreamingClientHandler.getClientProperties(connectorConfig);
+        if (this.registeredClientMap.containsKey(inputProps)) {
+          resultClient = this.registeredClientMap.get(inputProps);
+          LOGGER.debug("Using existing streaming client with name: {}", resultClient.getName());
+        } else {
+          Pair<Properties, SnowflakeStreamingIngestClient> propertiesAndClient =
+              this.streamingClientHandler.createClient(connectorConfig);
+          resultClient = propertiesAndClient.getSecond();
+          this.registeredClientMap.put(
+              propertiesAndClient.getFirst(), propertiesAndClient.getSecond());
+          LOGGER.debug("Created and registered new client with name: {}", resultClient.getName());
```

Comment:
should we print specific properties values here? so that we can debug which properties it is creating for and which not (in case things go wrong?)

Reply:
printing the non sensitive values here